File: Data Structures/#04.Linear Data Structures - Stack And Queue Exercises/Java/src/T06_Linked_Queue/LinkedQueue.java
package T06_Linked_Queue;

import org.w3c.dom.Node;

public class LinkedQueue<E> {

    private int size;
    private int elementsCount;
    private QueueNode<E> head;
    private QueueNode<E> tail;

    public int size() {
        return this.size;
    }

    private void setSize(int size) {
        this.size = size;
    }

    public void enqueue(E element) {

        if (this.elementsCount == 0) {
            QueueNode<E> node = new QueueNode<>();
            node.value = element;

            this.head = node;
            this.tail = node;
            this.elementsCount++;

            return;
        }

        QueueNode<E> newNode = new QueueNode<>();
        newNode.value = element;

        this.head.nextNode = newNode;
        newNode.prevNode = this.head;

        this.head = newNode;
        this.elementsCount++;
    }

    public E dequeue() {

        if(this.elementsCount == 0){
            throw new ArrayIndexOutOfBoundsException();
        }

        E element = this.tail.value;

        if(this.tail.nextNode != null){
            this.tail.nextNode.prevNode = null;
        }

        this.tail = this.tail.nextNode;

        this.elementsCount--;
        return element;
    }

    public E[] toArray() {
        E[] arr = (E[]) new Object[this.elementsCount];

        int indexCounter = 0;
        QueueNode currentNode = this.head;

        while (currentNode != null){
            arr[indexCounter] = (E) currentNode.value;
            currentNode = currentNode.prevNode;
            indexCounter++;

        }

        return arr;
    }


    private class QueueNode<E> {
        private E value;

        private QueueNode<E> nextNode;
        private QueueNode<E> prevNode;

        public E getValue() {
            return this.value;
        }

        private void setValue(E value) {
            this.value = value;
        }

        public QueueNode<E> getNextNode() {
            return this.nextNode;
        }

        public void setNextNode(QueueNode<E> nextNode) {
            this.nextNode = nextNode;
        }

        public QueueNode<E> getPrevNode() {
            return this.prevNode;
        }

        public void setPrevNode(QueueNode<E> prevNode) {
            this.prevNode = prevNode;
        }
    }
}

File: Data Structures/#04.Linear Data Structures - Stack And Queue Exercises/C#/T05_Linked_Queue/LinkedQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

public class LinkedQueue<T>
{
    private QueueNode<T> head;
    private QueueNode<T> tail;
    private int count;

    public int Count { get => count; set => count = value; }

    public void Enqueue(T element)
    {
        if (this.count == 0)
        {
            QueueNode<T> firstNode = new QueueNode<T>(element);
            this.head = firstNode;
            this.tail = firstNode;
            count++;
            return;
        }

        QueueNode<T> newNode = new QueueNode<T>(element);
        this.head.NextNode = newNode;
        newNode.PrevNode = this.head;
        this.head = newNode;
        count++;

    }
    public T Dequeue()
    {

        if (this.count == 0)
        {
            throw new InvalidOperationException();
        }

        T element = this.tail.Value;

        if (this.count == 1)
        {
            this.head = this.tail = null;
            count--;
            return element;
        }


        this.tail.NextNode.PrevNode = null;
        this.tail = this.tail.NextNode;
        count--;

        return element;
    }
    public T[] ToArray()
    {

        T[] result = new T[this.count];
        QueueNode<T> currentNode = this.tail;

        int counter = 0;
        while (currentNode != null)
        {
            result[counter] = currentNode.Value;
            currentNode = currentNode.NextNode;
            counter++;
        }

        return result;
    }

    private class QueueNode<T>
    {
        public T Value { get; private set; }

        public QueueNode<T> NextNode { get; set; }
        public QueueNode<T> PrevNode { get; set; }

        public QueueNode(T value, QueueNode<T> nextNode = null, QueueNode<T> prevNode = null)
        {
            Value = value;
            NextNode = nextNode;
            PrevNode = prevNode;
        }
    }
}
